Make interface prototype object creation imperative

tobie opened the pull request:
Closes #496.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/tobie/webidl/pull/494.html" title="Last updated on Apr 10, 2018, 3:07 PM GMT (eddf233)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/heycam/webidl/494/41bc728...tobie:eddf233.html" title="Last updated on Apr 10, 2018, 3:07 PM GMT (eddf233)">Diff</a>

TimothyGu commented:
Refs: https://github.com/heycam/webidl/issues/467

TimothyGu commented on index.bs:
```diff
@@ -10877,35 +10877,74 @@ defined on that interface.
+    1.  Otherwise, if |interface| is declared to inherit from another interface,
+        then set |proto| to the [=interface prototype object=] of the inherited interface.
+    1.  Otherwise, if |interface| is declared with the [{{LegacyArrayClass}}] [=extended attribute=],
+        then set |proto| to the {{%ArrayPrototype%}} of |realm|.
```
`|realm|.[[Intrinsics]].[[{{%ArrayPrototype%}}]]`

Check out https://tc39.github.io/ecma262/#table-21.

TimothyGu commented on index.bs:
```diff
@@ -10877,35 +10877,74 @@ defined on that interface.
+        then set |proto| to the [=interface prototype object=] of the inherited interface.
+    1.  Otherwise, if |interface| is declared with the [{{LegacyArrayClass}}] [=extended attribute=],
+        then set |proto| to the {{%ArrayPrototype%}} of |realm|.
+    1.  Otherwise, set |proto| to the {{%ObjectPrototype%}} of |realm|.
```
Ditto.

TimothyGu commented on index.bs:
```diff
@@ -10877,35 +10877,74 @@ defined on that interface.
+    1.  Otherwise, if |interface| is declared with the [{{LegacyArrayClass}}] [=extended attribute=],
+        then set |proto| to the {{%ArrayPrototype%}} of |realm|.
+    1.  Otherwise, set |proto| to the {{%ObjectPrototype%}} of |realm|.
+    1.  Let |interfaceProtoObj| be [=!=] <a abstract-op>ObjectCreate</a>(proto).
```
`|proto|`

TimothyGu commented on index.bs:
```diff
@@ -10877,35 +10877,74 @@ defined on that interface.
+    1.  If |interface| has any [=member=] declared with the [{{Unscopable}}] [=extended attribute=],
+        then:
+        1.  Set |unscopableObject| to be [=!=]
+            <a abstract-op>ObjectCreate</a>({{%ObjectPrototype%}} of |realm|).
```
Ditto.

TimothyGu commented on index.bs:
```diff
@@ -10877,35 +10877,74 @@ defined on that interface.
+            \[[Enumerable]]: <emu-val>false</emu-val>, \[[Configurable]]: <emu-val>true</emu-val>,
+            \[[Value]]: |unscopableObject|}.
+        1.  Perform [=!=] <a abstract-op>DefinePropertyOrThrow</a>(|interfaceProtoObj|, {{@@unscopables}}, |desc|).
+    1.  For each [=exposed=] [=regular attribute=] |attr| that is a [=member=] of |interface|,
```
Infra says this step should end with a colon instead of a comma.

TimothyGu replied:
["exposed"](https://heycam.github.io/webidl/#dfn-expose) right now seems to be only defined for interfaces, not members of interfaces.

TimothyGu commented on index.bs:
```diff
@@ -10877,35 +10877,74 @@ defined on that interface.
+        1.  Perform [=!=] <a abstract-op>DefinePropertyOrThrow</a>(|interfaceProtoObj|, {{@@unscopables}}, |desc|).
+    1.  For each [=exposed=] [=regular attribute=] |attr| that is a [=member=] of |interface|,
+        1.  Let |getter| be the result of creating an [=attribute getter=]
+            given |attr|, |interface|,  and |realm|.
```
nit: extra space before `and`. Ditto two lines down.

TimothyGu commented on index.bs:
```diff
@@ -10877,35 +10877,74 @@ defined on that interface.
+        and the string "<code>Prototype</code>".
+    1.  Let |desc| be the PropertyDescriptor{\[[Writable]]: <emu-val>false</emu-val>,
+        \[[Enumerable]]: <emu-val>false</emu-val>, \[[Configurable]]: <emu-val>false</emu-val>,
+        \[[Value]]: |value|}.
```
https://heycam.github.io/webidl/#issue-48167bcc says:

> Should define whether `@@toStringTag` is writable, enumerable and configurable. All `@@toStringTag` properties in the ECMAScript spec are non-writable and non-enumerable, and configurable.

I assume this "issue" has been resolved?

TimothyGu commented on index.bs:
```diff
@@ -10877,35 +10877,74 @@ defined on that interface.
+            Note: the algorithm to create an [=attribute setter=]
+            returns <emu-val>undefined</emu-val> if |attr| is [=read only=].
+        1.  Let |desc| be the PropertyDescriptor{\[[Get]]: |getter|, \[[Set]]: |setter|,
+            \[[Enumerable]]: <emu-val>true</emu-val>, \[[Configurable]]: <emu-val>true</emu-val>}.
```
This needs some more work based on https://heycam.github.io/webidl/#es-attributes, specifically about unforgeability. Same for operations.

tobie replied:
Thanks for catching that. Had completely missed it!

tobie commented:
> though operations from special declarations like iterable, maplike, and setlike still need to be defined. You probably already know this but spelling it out so that it doesn't get forgotten :)

🙈 🙉 🙊

TimothyGu commented on index.bs:
```diff
@@ -10877,35 +10874,74 @@ defined on that interface.
+        then set |proto| to the [=interface prototype object=] of the inherited interface.
+    1.  Otherwise, if |interface| is declared with the [{{LegacyArrayClass}}] [=extended attribute=],
+        then set |proto| to |realm|.\[[Intrinsics]].\[[{{%ArrayPrototype%}}]].
+    1.  Otherwise, set |proto| to |realm|.\[[Intrinsics]].\[[{{%ObjectPrototype%}}]]the {{%%}} of |realm|.
```
Looks like a regex fail :)

TimothyGu commented on index.bs:
```diff
@@ -10877,35 +10874,74 @@ defined on that interface.
+    1.  Otherwise, if |interface| is declared to inherit from another interface,
+        then set |proto| to the [=interface prototype object=] of the inherited interface.
+    1.  Otherwise, if |interface| is declared with the [{{LegacyArrayClass}}] [=extended attribute=],
+        then set |proto| to |realm|.\[[Intrinsics]].\[[{{%ArrayPrototype%}}]].
```
This becomes rendered as "<var>realm</var>.[[Intrinsics]].\[[%ArrayPrototype%]]"

TimothyGu commented on index.bs:
```diff
@@ -10877,35 +10874,74 @@ defined on that interface.
+    1.  If |interface| has any [=member=] declared with the [{{Unscopable}}] [=extended attribute=],
+        then:
+        1.  Set |unscopableObject| to be [=!=]
+            <a abstract-op>ObjectCreate</a>({{%ObjectPrototype%}} of |realm|).
```
Still needs conversion.

domenic commented on index.bs:
```diff
@@ -10877,35 +10877,54 @@ defined on that interface.
 These properties are described in more detail
 in sections [[#es-attributes]] and [[#es-operations]].
```
This paragraph was untouched, but the paragraphs about constants and the constructor property were deleted. Maybe best to delete them all, or alternately convert them into a brief non-normative summary (e.g. as a buletted list in a NOTE)?

domenic commented on index.bs:
```diff
@@ -10877,35 +10877,54 @@ defined on that interface.
+        then set |proto| to the [=named properties object=] of |interface|,
+        as defined in [[#named-properties-object]].
+    1.  Otherwise, if |interface| is declared to inherit from another interface,
+        then set |proto| to the [=interface prototype object=] of the inherited interface.
```
s/the inherited/that? Or s/inherited/inherited-from/?

domenic replied:
Also should state the interface prototype object in _realm_.

domenic commented on index.bs:
```diff
@@ -10877,35 +10877,54 @@ defined on that interface.
+        1.  Let |desc| be the PropertyDescriptor{\[[Value]]: |unscopableObject|,
+            \[[Writable]]: <emu-val>false</emu-val>, \[[Enumerable]]: <emu-val>false</emu-val>,
+            \[[Configurable]]: <emu-val>true</emu-val>}.
+        1.  Perform [=!=] <a abstract-op>DefinePropertyOrThrow</a>(|interfaceProtoObj|, {{@@unscopables}}, |desc|).
```
Style nit/matter of preference: I would install the property after filling it up with values, so move the following step up above these two.

domenic commented on index.bs:
```diff
@@ -10877,35 +10877,54 @@ defined on that interface.
+            1.  Perform [=!=] <a abstract-op>CreateDataProperty</a>(|unscopableObject|, |id|,
+                <emu-val>true</emu-val>).
+    1.  [=Expose the regular attributes=] of |interface| on |interfaceProtoObj| given |realm|.
+    1.  [=list/For each=] [=exposed=] [=regular operation=] |op| that is a [=member=] of |interface|:
```
Why did the other two get their own dedicated "expose the X"s, but not regular operations?

domenic commented on index.bs:
```diff
@@ -11117,48 +11118,82 @@ is the concatenation of the [=interface=]’s
-    *   |S| is the [=attribute setter=] created given the attribute, the interface,
-        and the [=relevant Realm=] of the object that is the location of the property.
+<div algorithm>
+    To <dfn>expose the regular attributes</dfn> of [=interface=] |interface| on |target|, given [=Realm=] |realm|, 
```
As above, _realm_ should be passed to "exposed" in all of these.

domenic commented on index.bs:
```diff
@@ -11117,48 +11118,82 @@ is the concatenation of the [=interface=]’s
+    To <dfn>expose the regular attributes</dfn> of [=interface=] |interface| on |target|, given [=Realm=] |realm|, 
+    run the following steps:
+
+    1.  Let |attribues| be the [=list=] of [=exposed=] [=regular attributes=] that are [=members=] of |interface|.
```
Typo "attribues" (missing t) in multiple places

domenic commented on index.bs:
```diff
@@ -11117,48 +11118,82 @@ is the concatenation of the [=interface=]’s
+    1.  [=Expose the attributes=] |attributes| of |interface| on |target| given |realm|.
+</div>
+
+<div algorithm>
```
This seems to be unused. I guess it could be done in the constructor ("interface object").

domenic commented on index.bs:
```diff
@@ -10877,35 +10877,54 @@ defined on that interface.
+            and <emu-val>true</emu-val> otherwise.
+        1.  Let |desc| be the PropertyDescriptor{\[[Value]]: |method|,
+            \[[Writable]]: |modifiable|, \[[Enumerable]]: <emu-val>true</emu-val>,
+            \[[Configurable]]: |modifiable|}.
```
This isn't correct for unforgeable operations :(. Those need to be on each instance, so putting them on the prototype is not right. They could be done in the constructor, i.e. https://heycam.github.io/webidl/#interface-object

TimothyGu replied:
Not all platform objects are created through the interface object though. We might need to extend https://heycam.github.io/webidl/#es-platform-objects instead.

domenic replied:
Ah right, so we can't really fix this until we've defined "create an object". So I guess in the meantime we should leave this declarative with a pointer to future work.

tobie replied:
Unforgeable regular operations never get that far. This is only for static unforgeable ones (and will be on the object itself for regular ops once we add a "create platform object" algorithm).

TimothyGu commented on index.bs:
```diff
@@ -10877,35 +10877,54 @@ defined on that interface.
+            \[[Enumerable]]: <emu-val>false</emu-val>, \[[Configurable]]: <emu-val>true</emu-val>,
+            \[[Value]]: |constructor|}.
+        1.  Perform [=!=] <a abstract-op>DefinePropertyOrThrow</a>(|interfaceProtoObj|, "<code>constructor</code>", |desc|).
+    1.  Return |interfaceProtoObj|.
```
This is still missing special operations and iterable/maplike/setlike declarations :slightly_frowning_face:

domenic replied:
It might be good to add a warning on what's missing, i.e. what features of the interfaceProtoObj are still installed declaratively, with pointers to the sections that do that. I definitely agree it's good to start with this though and not bite it all off at once.

tobie replied:
Fixed in 6b7ecd0b1ed658ba5260a14ad9b7c4588aa24d0a.

tobie commented on index.bs:
```diff
@@ -11117,48 +11118,82 @@ is the concatenation of the [=interface=]’s
-*   The value of the property is that which is obtained by [=converted to an ECMAScript value|converting=] the [=constant=]’s IDL value to an ECMAScript value.
-*   The property has attributes { \[[Writable]]: <emu-val>false</emu-val>, \[[Enumerable]]: <emu-val>true</emu-val>, \[[Configurable]]: <emu-val>false</emu-val> }.
+<div algorithm>
+    To <dfn>expose the constants</dfn> of [=interface=] |interface| on |target|,
```
oh, true. Not passing it explicitly to exposed yet, however.

TimothyGu commented on index.bs:
```diff
@@ -11117,48 +11115,83 @@ is the concatenation of the [=interface=]’s
+[=legacy callback interface objects=],
+[=interface prototype objects=], and
+on the single object that implements the interface,
+when an interface is declared with the [{{Global}}] [=extended attribute=].
```
Hmm. In [Chrome](https://cs.chromium.org/chromium/src/third_party/WebKit/Source/modules/filesystem/WindowFileSystem.idl):

```js
> Object.getOwnPropertyDescriptor(Window, 'PERSISTENT')
{value: 1, writable: false, enumerable: true, configurable: false}
> Object.getOwnPropertyDescriptor(Window.prototype, 'PERSISTENT')
{value: 1, writable: false, enumerable: true, configurable: false}
> Object.getOwnPropertyDescriptor(window, 'PERSISTENT')
undefined
```

This API is apparently nonstandard according to MDN so we don't have to worry about it much, but I haven't been able to find any other constants specified on interfaces declared to be [Global].

tobie replied:
Yeah, this is just rewording what already exists here.

domenic replied:
Might be worth a follow-up bug to disallow constants on [Global]s, as part of the general deprecation-ish of constants.

tobie replied:
Issue filed: https://github.com/heycam/webidl/issues/528.

tobie commented on index.bs:
```diff
@@ -10877,35 +10882,46 @@ defined on that interface.
-    \[[Prototype]] [=internal slot=] whose value is returned from
-    the following steps:
+    The [=interface prototype object=] for a given [=interface=] |interface| and [=Realm=] |realm|
+    is created as follows:
```
Yeah, there's a bunch of algorithms that need to be more linkable in there.

tobie replied:
Would you go for:

```
<h4>Named properties object</h4>

For every [=interface=] declared with the [{{Global}}] [=extended attribute=]
that [=support named properties|supports named properties=],
there must exist an object known as the [=named properties object=]
for that interface on which named properties are exposed.

<div algorithm>
    The <dfn export>named properties object</dfn>
    for a given [=interface=] |interface| and [=Realm=] |realm|, is created as follows:
    
    […]
</div>
```

or:

```
<h4>Named properties object</h4>

For every [=interface=] declared with the [{{Global}}] [=extended attribute=]
that [=support named properties|supports named properties=],
there must exist an object known as the <dfn export>named properties object</dfn>
for that interface on which named properties are exposed.

<div algorithm>    
    <dfn abstract-op>NamedPropertiesObjectCreate(|interface|, |realm|)</dfn>
   
    […]
</div>
```

TimothyGu replied:
I'd actually prefer

```html
<h4>Named properties object</h4>

For every [=interface=] declared with the [{{Global}}] [=extended attribute=]
that [=support named properties|supports named properties=],
there must exist an object known as the <dfn export>named properties object</dfn>
for that interface on which named properties are exposed.

<div algorithm>
    The [=named properties object=] for an [=interface=] |interface| in [=Realm=]
    |realm| is <dfn lt="create a named properties object">created</dfn> as follows:
    
    […]
</div>
```

then

```html
1.  Let |namedPropertiesObject| be the result of
    [=create a named properties object|creating a named properties object=] for
    |interface| in |realm|.
```

We don't use ES-style abstract operations much in Web IDL, and there are not many reasons to start now.

tobie replied:
WFM

tobie replied:
Doesn't work for Bikeshed, however: https://github.com/tabatkins/bikeshed/issues/1167

tobie commented on index.bs:
```diff
@@ -11289,45 +11322,57 @@ being thrown.  When not in strict mode, the assignment attempt will be ignored.
+    given [=Realm=] |realm| run the following steps:
+
+    1.  [=list/For each=] [=operation=] |op| of |operations|:
+        1.  If |op| has [=no allowable invocations=], then [=iteration/continue=].
```
I don't know. I'm keeping the original normative statements around here because I don't know better and want to keep the PR sort of manageable. That said, if this is something we can safely drop, I'm happy to do so.

tobie replied:
Removing as per #511.

tobie commented on index.bs:
```diff
@@ -3264,6 +3264,11 @@ and [=variadic=] operations
+An [=operation=], [=constructor=], or [=callback function=] whose
+[=effective overload set=] for a given [=interface=] with argument count 0
+is an [=set/is empty|empty=] [=set=],
+has <dfn>no allowable invocations</dfn>.
```
Mmm. So it seems we're making the approximation you describe in (1) across the spec (see for example in Namespace or operation bindings). Filed a separate issue for that here: #510.

tobie commented on index.bs:
```diff
@@ -11289,45 +11322,57 @@ being thrown.  When not in strict mode, the assignment attempt will be ignored.
+    To <dfn>expose the operations</dfn> |attributes| of [=interface=] |interface| on |target|
+    given [=Realm=] |realm| run the following steps:
+
+    1.  [=list/For each=] [=operation=] |op| of |operations|:
```
Removing this part altogether as per #511.

tobie commented on index.bs:
```diff
@@ -10877,35 +10882,46 @@ defined on that interface.
+        of that [=inherited interface=].
+    1.  Otherwise, if |interface| is declared with the [{{LegacyArrayClass}}] [=extended attribute=],
+        then set |proto| to |realm|.\[[Intrinsics]].[[{{%ArrayPrototype%}}]].
+    1.  Otherwise, set |proto| to |realm|.\[[Intrinsics]].[[{{%ObjectPrototype%}}]].
```
Feels like the pattern described in https://infra.spec.whatwg.org/#variables is the issue here. I mean you literally have an if…otherwise if…otherwise set of statements right above it, so there's absolutely no chance for this to happen, and yet it's not super obvious to the eye.

Should we write the below instead?

```
    1.  Let |proto| be |realm|.\[[Intrinsics]].[[{{%ObjectPrototype%}}]]..
    1.  If |interface| is declared with the [{{Global}}] [=extended attribute=],
        and |interface| [=support named properties|supports named properties=],
        then set |proto| to the [=named properties object=] of |interface|,
        as defined in [[#named-properties-object]].
    1.  Otherwise, if |interface| is declared to inherit from another interface,
        then set |proto| to the [=interface prototype object=] in |realm|
        of that [=inherited interface=].
    1.  Otherwise, if |interface| is declared with the [{{LegacyArrayClass}}] [=extended attribute=],
        then set |proto| to |realm|.\[[Intrinsics]].[[{{%ArrayPrototype%}}]].
```

tobie replied:
OK, so on second thought, and after having tried it out, your solution is better.

tobie commented on index.bs:
```diff
@@ -11117,48 +11118,82 @@ is the concatenation of the [=interface=]’s
-*   The value of the property is that which is obtained by [=converted to an ECMAScript value|converting=] the [=constant=]’s IDL value to an ECMAScript value.
-*   The property has attributes { \[[Writable]]: <emu-val>false</emu-val>, \[[Enumerable]]: <emu-val>true</emu-val>, \[[Configurable]]: <emu-val>false</emu-val> }.
+<div algorithm>
+    To <dfn>expose the constants</dfn> of [=interface=] |interface| on |target|,
```
"Define"?

tobie replied:
Went with "define" for now.

TimothyGu commented on index.bs:
```diff
@@ -12410,6 +12453,24 @@ attribute is the concatenation of the class string of the namespace, ".",
+
+<div algorithm>
+
+    A [=platform object=] implementing an [=interface=] |interface|, in [=Realm=] |realm|,
```
This needs to somehow say that *interface* is the [primary interface](https://heycam.github.io/webidl/#dfn-primary-interface), because a platform object can implement many different interfaces (i.e., inherited interfaces; so glad consequential interfaces are gone).

TimothyGu commented on index.bs:
```diff
@@ -12410,6 +12453,24 @@ attribute is the concatenation of the class string of the namespace, ".",
+
+    1.  Let |proto| be the [=interface prototype object=] for |interface| in |realm|.
+    1.  Let |obj| be [=ObjectCreate=](|proto|, |internalSlotsList|).
+    1.  Set |obj|.\[[SetPrototypeOf]] to the definition specified in [[#platform-object-setprototypeof]].
```
> ... the [[SetPrototypeOf]] internal method of a platform object *O* that **implements an interface with the [`Global`] extended attribute** is called ...

I admit I didn't see this condition until now either, but this PR should improve that.

Also this section needs to account for legacy platform objects as well.

1. Let *obj* be a newly created object.
2. If *interface* has a [`Global`] extended attribute, then set *obj*.[[SetPrototypeOf]] to ...
3. Otherwise, if *interface* supports indexed properties or named properties, then *obj* is a legacy platform object. Set *obj*'s internal methods to ... [those defined for legacy platform objects]

    *Right now, it's not very clear what happens if *interface* inherits from an interface that supports indexed/named properties. https://heycam.github.io/webidl/#idl-special-operations says:*

    > If an object implements more than one interface that defines a given special operation, then it is undefined which (if any) special operation is invoked for that operation.

    *so I assume it should support having inherited interfaces that have special operations. Yet, the definition for [support indexed properties](https://heycam.github.io/webidl/#dfn-support-indexed-properties), which [legacy platform object](https://heycam.github.io/webidl/#dfn-legacy-platform-object) uses, does not account for the possibility of inheriting a special operation. We should take care of this soon.*

4. If the specification defining *interface* provides an alternative definition for one or more internal methods for objects implementing *interface*, set *obj*'s internal methods accordingly.

5. Set all remaining internal methods of *obj* to the definitions specified in ECMA-262 section 9.1.

tobie replied:
(5) above is already taken care of by virtue of using `ObjectCreate`.

Not too sure what to do about your italicized section in (3).

tobie commented:
Hey folks, I think I mostly addressed the issues here, except all of the special ops work. I think it'll be easier to tackle special ops once we have the algorithm to create platform objects, which I've split out to separate branch.

Would really appreciate a review at this point to see if there are still unaddressed issues here. /cc @TimothyGu, @domenic.

domenic commented on index.bs:
```diff
@@ -10809,9 +10809,12 @@ the <code>typeof</code> operator will return "function" when applied to an inter
+    1.  Let |proto| be the [=interface prototype object=] of [=interface=] |I| in |realm|.
     1.  Perform [=!=] <a abstract-op>DefinePropertyOrThrow</a>(|F|, "<code>prototype</code>",
         PropertyDescriptor{\[[Value]]: |proto|, \[[Writable]]: <emu-val>false</emu-val>, \[[Enumerable]]: <emu-val>false</emu-val>, \[[Configurable]]: <emu-val>false</emu-val>}).
+    1.  [=Define the constants=] of [=interface=] |I| on |F| given |realm|.
```
I might name these "install", but define works too.

tobie replied:
Not addressing this as per your edit.

domenic commented on index.bs:
```diff
@@ -10867,41 +10872,50 @@ There must exist an <dfn id="dfn-interface-prototype-object" export>interface pr
+    1.  Otherwise, if |interface| is declared to inherit from another interface,
+        then set |proto| to the [=interface prototype object=] in |realm|
+        of that [=inherited interface=].
+    1.  Otherwise, if |interface| is the {{DOMException}} [=interface=],
```
Nice that we get to move this into here, instead of having it be an exception in the DOMException definition. We should remove the "must" from the custom bindings section, and instead make a statement of fact that refers back to this algorithm.

tobie replied:
Fixed in https://github.com/heycam/webidl/pull/494/commits/3c27f1d18ea6e8d248af369fc5ec9842ff084d35.

domenic commented on index.bs:
```diff
@@ -10867,41 +10872,50 @@ There must exist an <dfn id="dfn-interface-prototype-object" export>interface pr
+    1.  If |interface| has any [=member=] declared with the [{{Unscopable}}] [=extended attribute=],
+        then:
+        1.  Let |unscopableObject| be the result of performing [=!=]
+            <a abstract-op>ObjectCreate</a>(<emu-val>null</emu-val>).
```
The use of null here is a normative change that updates us to match the ES spec. Does it close an open bug? It'll be worth adding tests for :).

bzbarsky replied:
The null unscopable thing is https://github.com/heycam/webidl/issues/496

Safari already has the right behavior.  

I filed https://bugs.chromium.org/p/chromium/issues/detail?id=811029 (though of course bugs.chromium.org picked up the wrong account _again_, so if that can be changed or the issue deleted I would appreciate that) and https://bugzilla.mozilla.org/show_bug.cgi?id=1437255 which has idlharness tests for this stuff.

I can't file bugs on Edge, unfortunately, so someone else will have to do that.

tobie replied:
Filed the remaining bugs and consolidated all links to issues in https://github.com/heycam/webidl/issues/496#issuecomment-368589578.

domenic commented on index.bs:
```diff
@@ -10867,41 +10872,50 @@ There must exist an <dfn id="dfn-interface-prototype-object" export>interface pr
+        1.  Perform [=!=] <a abstract-op>DefinePropertyOrThrow</a>(|interfaceProtoObj|, "<code>constructor</code>", |desc|).
+    1.  Return |interfaceProtoObj|.
 </div>
 
```
We can imperative-ize the immutable prototype exotic objectness using the technique of this example: https://tc39.github.io/ecma262/#sec-boundfunctioncreate step 5. Actually, I see you used that same technique later on.

tobie replied:
You mean [[[SetPrototypeOf]]](https://heycam.github.io/webidl/#platform-object-setprototypeof), right? Doesn't that belong on the instance itself rather than on the proto, though?

tobie replied:
Fixed in 971cf61.

domenic commented on index.bs:
```diff
@@ -10766,11 +10766,11 @@ and is described in more detail in [[#interface-prototype-object]].
-    and in [=Realm=] |realm| is created as follows:
+    The [=interface object=] for a given non-callback [=interface=] |I|
+    with [=identifier=] |id| and in [=Realm=] |realm|
+    is <dfn lt="create an interface object">created</dfn> as follows:
```
Most of these algorithms don't return anything. Although I don't think they have callers yet, probably they should return the thing they create.

tobie replied:
Fixed in https://github.com/heycam/webidl/pull/494/commits/37ae0fb19dc2465ea20cb88adb746729b7dc6af7.

domenic commented on index.bs:
```diff
@@ -10983,42 +10978,46 @@ Note: Since a legacy callback interface object is a [=function object=]
+
+    1.  Let |proto| be null.
+    1.  If |interface| is declared to inherit from another interface,
+        then set |proto| to the [=interface prototype object=] in |realm| for the inherited interface.
```
Link inherited interface?

tobie replied:
Fixed in https://github.com/heycam/webidl/pull/494/commits/7aab506b677b0484740b146aa405cc0bb088d346.

domenic commented on index.bs:
```diff
@@ -10983,42 +10978,46 @@ Note: Since a legacy callback interface object is a [=function object=]
+    1.  Otherwise, set |proto| to |realm|.\[[Intrinsics]].[[{{%ObjectPrototype%}}]].
+    1.  Let |obj| be a newly created object.
+    1.  Set |obj|'s internal methods to the definitions specified in
+        [=ECMA-262 §9.1|ECMA-262 §9.1 Ordinary object internal methods and internal slots=], unless they are specified below.
```
Maybe s/below/the rest of this section, or even the rest of `[[#named-properties-object]]`?

tobie replied:
Fixed in https://github.com/heycam/webidl/pull/494/commits/727287a754551e961546524d7f5631d9191f26ef.

domenic commented on index.bs:
```diff
@@ -11113,48 +11108,85 @@ is the concatenation of the [=interface=]’s
-*   The value of the property is that which is obtained by [=converted to an ECMAScript value|converting=] the [=constant=]’s IDL value to an ECMAScript value.
-*   The property has attributes { \[[Writable]]: <emu-val>false</emu-val>, \[[Enumerable]]: <emu-val>true</emu-val>, \[[Configurable]]: <emu-val>false</emu-val> }.
+<div algorithm>
+    To <dfn>define the constants</dfn> of [=interface=] or [=namespace=] |definition| on |target|,
```
Namespaces can't have constants, so no need for that.

tobie replied:
Fixed in https://github.com/heycam/webidl/pull/494/commits/6d426d8fb24389a241e25b5f5ada0af35b890fe0.

domenic commented on index.bs:
```diff
@@ -11113,48 +11108,85 @@ is the concatenation of the [=interface=]’s
-        and the [=relevant Realm=] of the object that is the location of the property; and
-    *   |S| is the [=attribute setter=] created given the attribute, the interface,
-        and the [=relevant Realm=] of the object that is the location of the property.
+[=Static attributes=] are exposed on the [=interface object=].
```
I think here it's worth calling out that "define the unforgeable regular attributes" doesn't yet have any callers. Perhaps with a normative statement like "The implementation must perform define the unforgeable regular attributes on every instance created of every interface type. NOTE: we hope to turn this declarative into imperative later." (My phrasing is bad; end of long day; please phrase better :))

domenic replied:
I guess also another "must" for what to do with globals.

tobie replied:
Fixed in 6b7ecd0b1ed658ba5260a14ad9b7c4588aa24d0a.

domenic commented on index.bs:
```diff
@@ -11285,45 +11317,55 @@ being thrown.  When not in strict mode, the assignment attempt will be ignored.
-*   The value of the property is the result of [=creating an operation function=] given the operation,
-    the interface, and the [=relevant Realm=] of the object that is the location of the property.
+For each unique [=identifier=] of an [=exposed=] [=operation=] defined on the [=interface=],
+there must exist a corresponding property.
```
Here you kept the must, although eventually it's unnecessary (and you've been deleting it from previous sections, making these intros just statements of fact---which is good). Probably best to delete it from here too, although you'll need to---like the above---add some temporary musts for unforgeable and globals.

tobie replied:
Partially addressed in c6eb400

domenic commented on index.bs:
```diff
@@ -11285,45 +11317,55 @@ being thrown.  When not in strict mode, the assignment attempt will be ignored.
+            given |op|, |definition|, and |realm|.
+        1.  Let |modifiable| be <emu-val>false</emu-val> if |op| is [=unforgeable=]
+            and <emu-val>true</emu-val> otherwise.
+        1.  Let |id| be |op|'s [=identifier=].
```
Tiny nit: the order of this step and the next is switched here vs. for attributes. Might as well make them consistent.

tobie replied:
Fixed in 5585b02.

tobie commented:
OK, hopefully, I've addressed everything, now, and this is ready for a final review.

TimothyGu commented on index.bs:
```diff
@@ -10867,42 +10875,71 @@ There must exist an <dfn id="dfn-interface-prototype-object" export>interface pr
+        1.  Set the internal methods of |interfaceProtoObj|
+            which are specific to [=immutable prototype exotic objects=]
+            to the definitions specified in
+            [=ECMA-262 §9.4.7|ECMA-262 §9.4.7 Immutable prototype exotic objects=].
```
We already use section numbers in various places. We can follow this up with another PR, but I think for now consistency is good.

tobie replied:
Agree with @TimothyGu. Filed here: https://github.com/heycam/webidl/issues/542

tobie commented on index.bs:
```diff
@@ -10983,42 +10995,48 @@ Note: Since a legacy callback interface object is a [=function object=]
+    1.  Let |proto| be null.
+    1.  If |interface| is declared to inherit from another interface,
+        then set |proto| to the [=interface prototype object=] in |realm| for the [=inherited interface=].
+    1.  Otherwise, if |interface| is declared with the [{{LegacyArrayClass}}] [=extended attribute=],
```
Fixed.

tobie commented on index.bs:
```diff
@@ -12407,6 +12468,35 @@ attribute is the concatenation of the class string of the namespace, ".",
+as defined in [[#platform-object-setprototypeof]].
+
+For each [=interface=] |interface| implemented by a [=platform object=] |obj|,
+the following steps must be run:
```
Fixed.

tobie commented on index.bs:
```diff
@@ -12407,6 +12468,35 @@ attribute is the concatenation of the class string of the namespace, ".",
+1.  [=Define the regular attributes=] of |interface| on |obj|, given [=Realm=] |realm|.
+
+Additionaly, [=platform objects=] which implement an [=interface=]
+which has a [{{Global}}] [=extended attribute=],
```
Fixed.

tobie commented on index.bs:
```diff
@@ -10867,42 +10875,71 @@ There must exist an <dfn id="dfn-interface-prototype-object" export>interface pr
+        then:
+
+        Issue: Should an {{@@unscopables}} property also be defined if |interface| is
+        declared with the [{{Global}}] [=extended attribute=]?
```
Filed here: https://github.com/heycam/webidl/issues/544
